enhancements/installer/ipi-publishing-status.md

```diff
@@ -0,0 +1,206 @@
+  - [IR-302](https://issues.redhat.com/browse/IR-302)
+---
+
+# Expose Cluster Publishing Status
+
+## Summary
+
+IPI clusters can be configured to be public or private. This is done in the
+install-config.yaml using the `Publish` parameter. There is no current standard
+defined for how components should access this parameter to bootstrap themselves
+as public (External) or private (Internal).
+
+At the moment of writing, provisioning a private cluster means the cluster
+will be configured with a private DNS zone, internal (non internet facing)
+ingress, and internal API server.
+
+Exposing the cluster's publishing status empowers other components to bootstrap
+themselves as public or private according to the user request at install time.
+
+## Motivation
+
+A component should be able to know how to bootstrap itself according to user
+preference. A central place providing the cluster publishing status at install
+time makes more sense than the installer having to individually configure
+various components.
+
+### User Stories
+
+* As the author of the image registry operator, I want to easily discover the
+cluster publishing status, so that I can configure access the cloud storage objects
+accordingly.
+
+### Goals
+
+Expose the cluster publishing status used at install time in the Infrastructure
+object.
+
+### Non-Goals
+
+This enhancement does not cover support for updating the cluster publishing
+status on running clusters. This property is meant to reflect the `publish`
+parameter as it was set in the install-config.yaml.
+
+## Proposal
+
+The Infrastructure API is extended, adding the `Publish` field to the
+`InfrastructureStatus`.
+
+```go
+type InfrastructureStatus struct {
+	// ...
+
+	// publish indicates the publishing status of the cluster at install time.
+	// The default is 'External', which means the cluster will be publicly
+    // available on the internet.
+	// The 'Internal' mode indicates a private cluster, and operators should
+    // configure operands accordingly.
+    // +kubebuilder:default=External
+    // +kubebuilder:validation:Enum=Internal;External
+	Publish PublishMode `json:"publish"`
+}
+```
+
+To reflect the install-config publishing options, `PublishMode` will only
+support to valid options.
+
+```
+// PublishMode defines the publish mode of the cluster at install time.
+type PublishMode string
+
+const (
+	// "Internal" is for operators to configure their asserts as private
+    // when appropriate.
+	InternalPublishMode PublishMode = "Internal"
+
+	// "External" indicates that the cluster is publicly available on the internet.
+	ExternalPublishMode PublishMode = "External"
+)
+```
+
+### Workflow Description
+
+1. The cluster creator install a cluster via IPI, setting `publish: Internal` in installer-config.yaml
+2. The installer bootstraps the cluster, and sets `Publish: Internal` in the Infrastructure status
+3. Operators use this value to configure operands accordingly.
```

Comment:
How many operators do you expect will actually need to be aware of this configuration?

Reply:
I only expect registry operator to depend on this. We do have use cases for other parts of OCP that could move to this approach though. Namely ingress, DNS and apiserver. ATM I think they all rely on the installer to configure them as internal/external depending on what the user set in install-config.yaml.

Reply:
Does the registry operator have its own specific configuration object? I wonder if placing this globally within the infrastructure, but not using it globally could be confusing for users

Reply:
Yes the registry operator has its own configuration object. 
We (me + installer folks) have discussed the possibility of having the installer configure the registry to internal when appropriate, but a decentralized approach made more sense at the time.
I do agree that it could be confusing to support both, the installer way (the way it currently works) and this decentralized way. I think that ideally we would request for the teams being configured by the installer to consume this value instead, so that eventually only the decentralized approach is used.